### ISSUE TITLE
pki/ca/sign to allow CSR supplied values - to overcome not_before_duration limitation in sign-verbatim

### ISSUE THREAD

**Issue by sooslaca** [hashicorp/vault#10174]
Hi,

In #10104 I was told use sign instead of sign-verbatim. 
We try that but there are other 'issues' and still cannot achieve dating back `not_before `properly.

" It sounds like if the sign endpoint supported custom x509v3 extensions, there wouldn't be a need to use sign-verbatim in this case" <- correct, we don't need extension
But would need the ability to set CSR supplied O, E and L   (Org, Email, Locality) values.

I understand the purpose of /sign is not to trust user provided values, correct?
But I see https://www.vaultproject.io/api-docs/secret/pki#use_csr_common_name allows CN to be supplied but not other values.

1) would it be possible to have flag for other values as well to be used?
2) would it be possible in sign-verbatim to use `notbefore` as defined in the role?
3) any other solution by which we can achieve the desired certificate to be signed?

Thanks

**Comment by sooslaca** [comment]
To make it more precise reg sign-verbatim, previously I was referred to @jefferai's comment which is:
"Yes, the point of sign-verbatim is that you are taking the values verbatim from the CSR rather than applying role parameters to it."

Which is fine.
Quote: @Naugrimm "The problem is, that notBefore- and notAfter-date of a certificate can not be taken from the CSR, as they cannot be specified"
Eg: in CSR there cannot be such values defined.

Exactly the same problem we ran into, hence suggesting to let be able to define `notbefore` as a json attribute in the request for `sign-verbatim`.

Thanks

**Comment by cipherboy** [comment]
@Gabrielopesantos I re-read this issue and point number two (copying `not_before_duration` from the role) might be a good idea, since [we already copy other relevant TTL fields there](https://github.com/hashicorp/vault/blob/main/builtin/logical/pki/path_issue_sign.go#L205-L211).

This one might be a little easier than the other one. :-)

**Comment by Gabrielopesantos** [comment]
With point two you mean this one "would it be possible in sign-verbatim to use `notbefore` as defined in the role?", right?
I agree, does not seem as complex as the other. I'll start with this one.

**Comment by cipherboy** [comment]
@Gabrielopesantos Yep! And thank you! Let me know if you have any questions.

**Comment by Gabrielopesantos** [comment]
Hey, investigated a bit and, am I missing something or this is just adding an `if` statement for the `NotBeforeDuration` field just like the `TTL` or `MaxTTL`?

**Comment by cipherboy** [comment]
Nope, you're not missing anything that I know of. :-) Sorry, I don't seem to have much in the spectrum between "if statement" and "write a mildly complicated new feature"...

**Comment by Gabrielopesantos** [comment]
Right, ran the tests and this addition does break anything but I guess it still should be tested, right? If yes, I suppose the test should also be created in `backend_test.go`?

**Comment by cipherboy** [comment]
@Gabrielopesantos Yeah, I wouldn't bother adding a new one, but there's an existing one around line 2179 (`create a role entry; we use this to check that sign-verbatim when used with a role is still honoring TTLs`) that can be updated to set the `not_before_duration` to something different and test it.

**Comment by cipherboy** [comment]
@sooslaca `sign-verbatim` now allows the use of `not_before_duration` from the role thanks to @Gabrielopesantos. I believe this should address this issue. Feel free to reopen if it doesn't :-)
